Fix python runtime deadlock due to multithreading + fork

varunbpatil opened the pull request:
Drop support for Python2.7 runtime because it has reached End Of Life
and also because "spawn" and "forkserver" multiprocessing contexts are
not supported in Python 2.7 which are required to fix the deadlock.

Fixes #94

varunbpatil commented:
Hi @andresmgot , I have tested this change with 1000 requests and a concurrency of 50. I did not observe any deadlock. Without this change, the same command results in a deadlock pretty easily. The function I used to test this is `stable/python/examples/preload.py` which I have included in this PR.

```bash
$ hey -n 1000 -c 50 -m GET -t 0 -disable-keepalive http://localhost:8080
                    
Summary:                                                 
  Total:        7.1151 secs
  Slowest:      2.9556 secs
  Fastest:      0.0102 secs         
  Average:      0.2326 secs                          
  Requests/sec: 140.5466                             
                                                     
  Total data:   11000 bytes                          
  Size/request: 11 bytes                             
                 
Response time histogram: 
  0.010 [1]     |     
  0.305 [939]   |________________________________________
  0.599 [4]     |   
  0.894 [0]     |   
  1.188 [21]    |_  
  1.483 [15]    |_  
  1.777 [8]     |   
  2.072 [5]     |   
  2.367 [1]     |   
  2.661 [0]     |
  2.956 [6]     |                   
                                                     
                                                     
Latency distribution:                                
  10% in 0.0409 secs                                 
  25% in 0.0946 secs                                 
  50% in 0.1819 secs
  75% in 0.2225 secs     
  90% in 0.2433 secs  
  95% in 1.1023 secs
  99% in 2.0323 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0582 secs, 0.0102 secs, 2.9556 secs
  DNS-lookup:   0.0007 secs, 0.0000 secs, 0.0132 secs
  req write:    0.0001 secs, 0.0000 secs, 0.0074 secs
  resp wait:    0.1724 secs, 0.0099 secs, 1.9005 secs
  resp read:    0.0019 secs, 0.0000 secs, 0.1229 secs

Status code distribution:
  [200] 1000 responses
```

varunbpatil commented on .circleci/config.yml:
```diff
@@ -10,6 +10,7 @@ defaults: &defaults
 run_test: &run_test
   steps:
     - checkout
+    - run: echo "export GIT_SHA1=${CIRCLE_SHA1}" >> $BASH_ENV
```
I'm using this environment variable in `stable/python/Makefile` to refer to files from the pull-request rather than fixed files from `master` branch.

varunbpatil commented on stable/python/Dockerfile.3.6:
```diff
@@ -17,9 +17,9 @@ RUN python ./get-pip.py
 
 WORKDIR /
-ADD kubeless.py .
+ADD _kubeless.py .
```
Renaming this file to something that has a very low likelihood of name collision with a user-defined module.
For example, the user can provide `--from-file kubeless.py --handler kubeless.handler` and it would still work.
This wasn't required when we were using `fork`, but is required when we use `forkserver`.

varunbpatil commented on stable/python/_kubeless.py:
```diff
@@ -1,19 +1,22 @@
 
+import importlib
 import os
-import imp
```
The `imp` module is deprecated since python 3.4 in favor of `importlib`.

varunbpatil commented on stable/python/_kubeless.py:
```diff
@@ -71,16 +74,13 @@ def handler():
         'event-type': req.get_header('event-type'),
         'event-time': req.get_header('event-time'),
         'event-namespace': req.get_header('event-namespace'),
-        'extensions': {
```
The `req` object is not picklable. Therefore, it cannot be passed to a child process when using `spawn` or `forkserver`. 

This used to work with `fork` because nothing is actually pickled and sent to the child process when using `fork`. The child process used to share memory with the parent process and therefore had access to everything in the parent directly.

varunbpatil replied:
I will have to check on that and get back to you. Will probably have to selectively discard some components in the `req` object which are not picklable (i.e, serializable).

andresmgot replied:
:+1: as far as I can tell, the most important thing is to keep request headers

varunbpatil replied:
I have added a new class `PicklableBottleRequest` to convert the original request object into a request object that can be pickled (serialized). It does this by removing components that cannot be pickled.

**Components that are removed from the original request object:**
1. `wsgi.errors` - A file-like object that the Bottle app writes errors to after the request is complete. I don't see any use for this within the request itself.
2. `wsgi.input` - A file-like object that contains input data, forms, files, etc. This can be pickled as long as the length of the data is less than `MEMFILE_MAX` in which case Bottle uses in-memory buffers (`io.BytesIO`). If the length is greater than `MEMFILE_MAX`, Bottle uses `tempfile.TemporaryFile` which cannot be pickled. So, leaving this out altogether for now.
3. `bottle.app` - A reference to the Bottle application (`bottle.app()`) itself. I don't know why this would be useful.
4. `bottle.route` and `route.handle` - Both are reference to the `bottle.Route` object for the current request. In our case, we only have one route `/`. So, I don't see any use for this.

**Components that are retained:**
All custom headers and URL arguments (`?abc=X&def=Y`) along with all the convinience methods like `get_header()`. I have added an example of using these in `examples/request.py`.

varunbpatil replied:
I am checking if it is possible to pickle the request body as well. If that can be done, the user code can access multipart form data as well.

varunbpatil commented on stable/python/_kubeless.py:
```diff
@@ -97,18 +97,63 @@ def handler():
                 return res
 
 
+def preload():
```
When using `forkserver`, the forkserver is not started until the first process is created. I'm using this function to create that first process before the cherrypy server is even started so that everything is primed and ready to go as soon as the web server is up and we start receiving requests.

varunbpatil commented on stable/python/_kubeless.py:
```diff
@@ -97,18 +97,63 @@ def handler():
+    import multiprocessing as mp
     import requestlogger
+
+    mp_context = os.getenv('MP_CONTEXT', 'forkserver')
```
Using `forkserver` by default because it is faster than `spawn` and also because it allows us to preload data that can be used by all requests as shown in `stable/python/examples/preload.py`.

If `forkserver` is not available, we get back a `ValueError` from `mp.get_context()` in which case we will default to `spawn`.

andresmgot commented on stable/python/_kubeless.py:
```diff
@@ -97,18 +97,63 @@ def handler():
+            d.start()
+            d.join()
+
+    except ValueError:
+        # Default to 'spawn' if 'forkserver' is unavailable.
+        ctx = mp.get_context('spawn')
+        logging.warn(
+            f'"{mp_context}" multiprocessing context is unavailable. Using "spawn"'
+        )
```
what would be the use case in which is better to use the `spawn`ed process?

varunbpatil replied:
None actually. The only reason I've included it is because not all platforms have support for `forkserver` yet.
Also, `spawn` is very safe and almost guaranteed to not have any problems and is available on all platforms which could be a quick workaround in case somebody faces an issue with `forkserver` (they just have to set an environment variable to switch to `spawn`).

`forkserver` is there for superior performance provided it is used properly. For example, If I were to use `spawn` with the example file I've included in the PR (`preload.py`), each and every request will take 20 seconds to complete because 20 seconds is how long it takes to load the two (hypothetical) ML models in that example (10 seconds to load one ML model). When I use `forkserver`, those models are only loaded once and that memory is shared with all the child processes, so each request only takes a few milliseconds to complete.

andresmgot replied:
sounds good

andresmgot reviewed (commented):
Thank you for the PR and the notes @varunbpatil! I just have a request for maintaining the `req` object and maintain backward compatibility.

varunbpatil commented on stable/python/_kubeless.py:
```diff
@@ -33,7 +36,7 @@
 
 function_context = {
-    'function-name': func,
+    'function-name': func.__name__,
```
`func` is actually the function object itself and not the function name as the dict key suggests.
`func.__name__` gets us the name of the function which should be the same as the `FUNC_HANDLER` environment variable.

I just checked the golang runtime and there we are passing the `FUNC_HANDLER` environment variable itself.
https://github.com/kubeless/runtimes/blob/60a58bf34e1544dd5b33e3db3fbe910b568666fe/stable/golang/server/kubeless.go.tpl#L41-L44

varunbpatil replied:
The difference between `func` and `func.__name__` is shown below (from the python interpreter).

```python
>>> def func(): pass
...
>>>
>>>
>>> type(func)
<class 'function'>
>>> func
<function func at 0x7f0255132730>
>>>
>>>
>>> type(func.__name__)
<class 'str'>
>>> func.__name__
'func'
```

varunbpatil commented on stable/python/Makefile:
```diff
@@ -131,9 +120,17 @@ post-python-verify:
+	kubeless function deploy python-preload --runtime python3.6 --handler preload.handler --from-file examples/preload.zip --timeout 3
+
+python-preload-verify:
+	kubectl rollout status deployment/python-preload && sleep 60
```
I've deliberately added two sleep's of 10 seconds each in the example file - `examples/preload.py`- to simulate an operation that takes long to complete and show the performance benefit of `forkserver` compared to `spawn` (each request takes few milliseconds for `forkserver` vs 20 seconds for `spawn`).

However, before we start serving requests, `preload.py` gets loaded twice.
1. It gets loaded first time when we do `mod = importlib.import_module(os.getenv('MOD_NAME'))`
2. It gets loaded a second time inside the forkserver when the forkserver is started - `d = ctx.Process(target=preload)`

So, the total delay before we can start serving requests is 2 * (10 + 10) = 40 seconds. 
I just added 20 seconds on top of that as a safety net.

I can reduce the sleep duration in `preload.py` to 4 seconds (2 + 2 for each of the ML models) and reduce the sleep in this Makefile to 2 * (2 + 2) = 8 + 2 (safety net) = 10 seconds.

Ideally we would need to add a `readinessProbe` so that `kubectl rollout status` will tell us exactly when the pod is ready to serve requests.

andresmgot replied:
> I can reduce the sleep duration in `preload.py` to 4 seconds (2 + 2 for each of the ML models) and reduce the sleep in this Makefile to 2 * (2 + 2) = 8 + 2 (safety net) = 10 seconds.

Yes, that'd be better.

varunbpatil replied:
I have made the change.

varunbpatil commented:
> Thank you for the PR and the notes @varunbpatil! I just have a request for maintaining the `req` object and maintain backward compatibility.

Hi @andresmgot, sorry for the delay in fixing this. I have made changes to preserve the `request` object so that custom headers, query strings, forms, file uploads, etc can be accessed in the user code. The changes work, but, it is starting to look like a hack which requires understanding of the internals of the Bottle framework. I can't think of any other way to make the request object serializable.

Maybe a better design would be to run multiple single-threaded servers behind a reverse-proxy to handle multiple requests in parallel (i.e, a pre-fork model). This comes with its own problems for handling timeouts, managing multiple server processes, synchronized logging, etc.

I am thinking this PR should be sufficient to fix this somewhat critical bug while maintaining backward compatibility. Maybe we can re-evaluate the design in the near future.

I would be glad to hear your thoughts on this.

andresmgot reviewed (approved):
Thanks @varunbpatil, I don't have a better suggestion for this so let's go with it and see if people find any issue. Since you have contributed a lot to this runtime, do you mind adding yourself to the `OWNERS` file? That way you can help me with future issues/feature requests of the runtime :)

andresmgot commented:
it seems that the test `get-python-deps` is failing with the new image, can you check it @varunbpatil ?

varunbpatil commented:
> it seems that the test `get-python-deps` is failing with the new image, can you check it @varunbpatil ?

Will check. Seems like something to do with Prometheus.

> do you mind adding yourself to the OWNERS file?

Will do.

varunbpatil commented:
> > it seems that the test `get-python-deps` is failing with the new image, can you check it @varunbpatil ?
> 
> Will check. Seems like something to do with Prometheus.
> 
> > do you mind adding yourself to the OWNERS file?
> 
> Will do.

Hi @andresmgot, I have fixed the issue. Those prometheus metrics variables needed to be moved inside `if __name__ == '__main__'` because they can only be created once. I missed it earlier since I had commented out some of the code. I have also added myself to the `OWNERS` file.

I have force-pushed my changes, so your changes to the image sha256 will have to be done again since the image will be different. Sorry for the trouble.

varunbpatil commented:
Hi @andresmgot , not sure what the current failure is. I tried with the new Docker images on a local Kubernetes cluster and it works for me.

```bash
[~/runtimes/stable/python] $ make get-python-deps
cd examples && zip hellowithdeps.zip hellowithdeps.py hellowithdepshelper.py
/home/varun.patil/runtimes/stable/python/examples
updating: hellowithdeps.py (stored 0%)
updating: hellowithdepshelper.py (deflated 29%)
kubeless function deploy get-python-deps --runtime python3.6 --handler hellowithdeps.foo --from-file examples/hellowithdeps.zip --dependencies examples/requirements.txt --image-pull-policy IfNotPresent
INFO[0000] Deploying function...
INFO[0000] Function get-python-deps submitted for deployment
INFO[0000] Check the deployment status executing 'kubeless function ls get-python-deps'

[~/runtimes/stable/python] $ make get-python-deps-verify
kubectl rollout status deployment/get-python-deps && sleep 2
deployment "get-python-deps" successfully rolled out
kubeless function call get-python-deps |egrep Google
Google
```

Is there a way to get the deployment status?

andresmgot commented:
> Is there a way to get the deployment status?

Let me check, I am going to run the tests with SSH and report back the issue.

andresmgot commented:
so it's working properly, it seems that even though the pod is running, it's not able to accept requests right away. Note that there is no error in the logs so the request is not even handled:

```
circleci@default-dca24841-1afb-473b-90f7-86c13f586b8c:~/project/stable/python$ make get-python-deps-verify 
kubectl rollout status deployment/get-python-deps && sleep 2
deployment "get-python-deps" successfully rolled out
kubeless function call get-python-deps |egrep Google
Google

circleci@default-dca24841-1afb-473b-90f7-86c13f586b8c:~/project/stable/python$ kubectl logs get-python-deps-6b694c6f98-2hn6d
Bottle v0.12.13 server starting up (using CherryPyServer(numthreads=10))...
Listening on http://0.0.0.0:8080/
Hit Ctrl-C to quit.

172.17.0.1 - - [28/Sep/2020:09:05:26 +0000] "GET /healthz HTTP/1.1" 200 2 "" "kube-probe/1.15" 0/237
172.17.0.1 - - [28/Sep/2020:09:05:56 +0000] "GET /healthz HTTP/1.1" 200 2 "" "kube-probe/1.15" 0/139
172.17.0.1 - - [28/Sep/2020:09:06:26 +0000] "GET /healthz HTTP/1.1" 200 2 "" "kube-probe/1.15" 0/142
172.17.0.1 - - [28/Sep/2020:09:06:30 +0000] "GET / HTTP/1.1" 200 6 "https://10.142.4.173:8443/api/v1/namespaces/default/services/get-python-deps:http-function-port/proxy" "kubeless/v0.0.0 (linux/amd64) kubernetes/$Format" 0/80126
```

Adding a small sleep should be enough.

varunbpatil commented:
But, there is already a 2 second sleep for every function. Do you want me to increase it for every function or just for the first function(hopefully others will be ready to accept requests by then)?

> By the way, pull before pushing more changes so the change in the jsonnet doesn't get lost.

Will do.

I'll also add a dump of the K8s objects on failure similar to https://github.com/kubeless/kubeless/blob/25d6376c91730f0fecb4c6f318b368dc4ce4eb03/script/integration-tests#L86-L104

andresmgot commented:
> But, there is already a 2 second sleep for every function. Do you want me to increase it for every function or just for the first function(hopefully others will be ready to accept requests by then)?

just for that test should be fine

varunbpatil commented:
> > But, there is already a 2 second sleep for every function. Do you want me to increase it for every function or just for the first function(hopefully others will be ready to accept requests by then)?
> 
> just for that test should be fine

Hi @andresmgot, increased the sleep duration to 5s for just the first test. CI is passing.

andresmgot commented:
Great, thanks!
